pkg/codegen/internal/test/testdata/output-funcs-tfbridge20/dotnet/GetAmiIds.cs

```diff
@@ -0,0 +1,194 @@
+        public List<string> ExecutableUsers
+        {
+            get => _executableUsers ?? (_executableUsers = new List<string>());
+            set => _executableUsers = value;
+        }
+
+        [Input("filters")]
+        private List<Inputs.GetAmiIdsFilterArgs>? _filters;
+
+        /// <summary>
+        /// One or more name/value pairs to filter off of. There
+        /// are several valid keys, for a full reference, check out
+        /// [describe-images in the AWS CLI reference][1].
+        /// </summary>
+        public List<Inputs.GetAmiIdsFilterArgs> Filters
+        {
+            get => _filters ?? (_filters = new List<Inputs.GetAmiIdsFilterArgs>());
+            set => _filters = value;
+        }
+
+        /// <summary>
+        /// A regex string to apply to the AMI list returned
+        /// by AWS. This allows more advanced filtering not supported from the AWS API.
+        /// This filtering is done locally on what AWS returns, and could have a performance
+        /// impact if the result is large. It is recommended to combine this with other
+        /// options to narrow down the list AWS returns.
+        /// </summary>
+        [Input("nameRegex")]
+        public string? NameRegex { get; set; }
+
+        [Input("owners", required: true)]
+        private List<string>? _owners;
+
+        /// <summary>
+        /// List of AMI owners to limit search. At least 1 value must be specified. Valid values: an AWS account ID, `self` (the current account), or an AWS owner alias (e.g. `amazon`, `aws-marketplace`, `microsoft`).
+        /// </summary>
+        public List<string> Owners
+        {
+            get => _owners ?? (_owners = new List<string>());
+            set => _owners = value;
+        }
+
+        /// <summary>
+        /// Used to sort AMIs by creation time.
+        /// </summary>
+        [Input("sortAscending")]
+        public bool? SortAscending { get; set; }
+
+        public GetAmiIdsArgs()
+        {
+        }
+    }
+
+    public sealed class GetAmiIdsInvokeArgs : Pulumi.InvokeArgs
+    {
+        [Input("executableUsers")]
+        private InputList<string>? _executableUsers;
+
+        /// <summary>
+        /// Limit search to users with *explicit* launch
+        /// permission on  the image. Valid items are the numeric account ID or `self`.
+        /// </summary>
+        public InputList<string> ExecutableUsers
+        {
+            get => _executableUsers ?? (_executableUsers = new InputList<string>());
+            set => _executableUsers = value;
+        }
+
+        [Input("filters")]
+        private InputList<Inputs.GetAmiIdsFilterInputArgs>? _filters;
+
+        /// <summary>
+        /// One or more name/value pairs to filter off of. There
+        /// are several valid keys, for a full reference, check out
+        /// [describe-images in the AWS CLI reference][1].
+        /// </summary>
+        public InputList<Inputs.GetAmiIdsFilterInputArgs> Filters
```

Comment:
@justinvp this is the pointer to the counter-example where Apply approach broke down.

Using terminology from the doc by @pgavlin : https://github.com/pulumi/pulumi/blob/master/developer-docs/architecture/type-system.md#inputshapet

Suppose we have ideal type X, and we have 

- `dotnetType(X) = GetAmiIdsArgs`
-  then `dotnetType(inputShape(X)) = GetAmiIdsInvokeArgs`
- But as part of covering the fields of this record, we discover `dotnetType(Y) = Inputs.GetAmiIdsFilterArgs` 
- and `dotnetType(inputShape(Y)) = Inputs.GetAmiIdsFilterInputArgs`
- `X.Filters` is of type type `List<Y>`

Now with the Apply approach, we need to build a .NET function `Inputs.GetAmiIdsInvokeArgs -> Output<GetAmiIdsArgs>`. The previous implementation was not cutting it. It stumbled on auto-converting `Inputs.GetAmiIdsFilterInputArgs` to `Inputs.GetAmiIdsFilterArgs`.

Reply:
I'd continue arguing that codegen or else SDK needs to reliably be able to do conversions of the form `inputShape(X) -> Input<X>`  in the runtime of the target language, of which the above is a special case :) However poking at the existing SDK it did not sound trivial. I'd have to be reflection, and I don't think it's built yet. Then I noticed that serde has this capability (sort of) and changed the approach.